android/app/src/main/java/com/zulipmobile/notifications/FCMPushNotifications.kt

```diff
@@ -69,124 +76,215 @@ internal fun onReceived(context: Context, conversations: ConversationMap, mapDat
+            fcmMessage.recipient.pmUsers.size - 2,
+            fcmMessage.sender.fullName,
+            fcmMessage.recipient.pmUsers.size - 2
+        )
+        is Recipient.Pm -> fcmMessage.sender.fullName
+    }
+    val isGroupConversation = when (fcmMessage.recipient) {
+        is Recipient.Stream -> true
+        is Recipient.GroupPm -> true
+        is Recipient.Pm -> false
+    }
+    val groupKey = extractGroupKey(fcmMessage.identity)
+    val conversationKey = extractConversationKey(fcmMessage)
+    var (notificationId, notification) = getActiveNotification(context, conversationKey)
+    notificationId = notificationId?: fcmMessage.zulipMessageId
 
-    builder.setDefaults(NotificationCompat.DEFAULT_VIBRATE or NotificationCompat.DEFAULT_LIGHTS)
+    val messagingStyle = when (notification) {
+        null -> NotificationCompat.MessagingStyle(selfUser)
+        else -> NotificationCompat.MessagingStyle
+            .extractMessagingStyleFromNotification(notification)
+            ?: NotificationCompat.MessagingStyle(selfUser)
+    }
+    messagingStyle
+        .setConversationTitle(title)
+        .setGroupConversation(isGroupConversation)
+        .addMessage(fcmMessage.content, fcmMessage.timeMs, sender)
 
-    val dismissIntent = Intent(context, NotificationIntentService::class.java)
-    dismissIntent.action = ACTION_CLEAR
-    val piDismiss = PendingIntent.getService(context, 0, dismissIntent, 0)
-    val action = NotificationCompat.Action(android.R.drawable.ic_menu_close_clear_cancel, "Clear", piDismiss)
-    builder.addAction(action)
+    val messageCount = messagingStyle.messages.size
 
-    val soundUri = getNotificationSoundUri(context)
-    builder.setSound(soundUri)
-    return builder
+    val builder = NotificationCompat.Builder(context, CHANNEL_ID)
+    builder.apply {
+        // This should agree with `BRAND_COLOR` in the JS code.
+        color = Color.rgb(100, 146, 254)
+        if (BuildConfig.DEBUG) {
+            setSmallIcon(R.mipmap.ic_launcher)
+        } else {
+            setSmallIcon(R.drawable.zulip_notification)
+        }
+        setAutoCancel(true)
+        setStyle(messagingStyle)
+        setGroup(groupKey)
+        setSound(getNotificationSoundUri())
+        setContentIntent(createViewPendingIntent(fcmMessage, context))
+        setNumber(messageCount)
+
+        val extraData = Bundle()
+        extraData.putInt("zulipMessageId", fcmMessage.zulipMessageId)
+        extras = extraData
+    }
+
+    val summaryNotification = createSummaryNotification(context, fcmMessage, groupKey)
+
+    NotificationManagerCompat.from(context).apply {
+        // here notifications are identified by TAG + notificationId
+        // simply using notificationId is not sufficient and will
+        // fail in case of different account in same realm.
+        notify(groupKey, fcmMessage.identity.realmId, summaryNotification.build())
+        notify(conversationKey, notificationId, builder.build())
```

Comment:
Looking again at these notification "id"s, the second argument to each of these `notify` calls:

Do we ever use the fact that the summaries have the realm ID as their notification ID? I think we don't; and I think if we did, that would probably be a bug, because realm IDs are per-server and in fact are highly likely to collide between any two servers other than Zulip Cloud (because they'll probably both be the only realm on the server.) If we don't, how about just using a constant ID, like 1? That would help make it clear that we're relying entirely on the tags.

Similarly, do we ever use the fact that the notification IDs for individual conversations vary, or that they're Zulip message IDs? I think we don't, in which case it'd similarly be good to just use a constant. It can even be the same constant -- we've designed these keys so they don't collide between the two types, either.

Reply:
using NOTIFICATION_ID, like the main branch here.